File: 2022/10/10.py

def generator(input) : 
    reg=[]
    v = 1
    for i in input.splitlines() : 
        reg.append(v)
        if "addx" in i : 
            reg.append(v)
            v += int(i.split(' ')[1])
    return reg
            
            
def part_1(input) : 
    return sum([x * input[x-1] for x in [20, 60, 100, 140, 180, 220]])

def part_2(input) :
    dsp = [" ", "■"]
    crt = [dsp[ i % 40 in [v-1 , v, v+1]]+((i % 40) == 39)*'\n' for i, v in enumerate(input)]
    return "".join(crt)[:-1]

File: 2022/10/10.rs
aoc2022::main!(); 

fn generator(input : &str) -> Vec<i32> {
    input
        .lines()
        .map(|l| if l == "noop" {(1, 0)} else {(2 , l.split(' ').collect_vec()[1].parse().unwrap())})
        .fold((Vec::new() ,1) , |(c , v) , (n, inc)| {
            ([c , repeat_n(v, n).collect_vec()].concat() , v + inc)
        }
    ).0
}

fn part_1(ops : Vec<i32>) -> i32 {
    ops
        .iter()
        .enumerate()
        .filter(|(n , _)| ((*n as i32) - 19 ) % 40 == 0 )
        .fold(0, |acc, (n , v)| acc + (1 + n as i32) * v)
}

fn part_2(ops : Vec<i32>) -> String {
    ops
        .iter()
        .enumerate()
        .map(|(n ,&v)| [v-1 , v, v+1].contains(&((n as i32) % 40)))
        .map(|b| if b {'■'} else{' '})
        .chunks(40)
        .into_iter()
        .map(|c| c.collect::<String>())
        .join("\n")
}